Make S3 custom query parameter optional

DaveCTurner opened the pull request:
Today Elasticsearch will record the purpose for each request to S3 using
a custom query parameter[^1]. This isn't believed to be necessary outside of
the ECH/ECE/ECK/... managed services, and it adds rather a lot to the
request logs, so with this commit we make the feature optional and
disabled by default.

[^1]: https://docs.aws.amazon.com/AmazonS3/latest/userguide/LogFormat.html#LogFormatCustom

elasticsearchmachine commented:
Hi @DaveCTurner, I've created a changelog YAML for you. *Note* that since this PR is labelled `>breaking`, you need to update the changelog YAML to fill out the extended information sections.

elasticsearchmachine commented:
Hi @DaveCTurner, I've updated the changelog YAML for you. *Note* that since this PR is labelled `>breaking`, you need to update the changelog YAML to fill out the extended information sections.

DaveCTurner committed:
Make S3 custom query parameter optional

Today Elasticsearch will record the purpose for each request to S3 using
a custom query parameter. This isn't believed to be necessary outside of
the ECH/ECE/ECK/... managed services, and it adds rather a lot to the
request logs, so with this commit we make the feature optional and
disabled by default.

DaveCTurner committed:
CI poke

DaveCTurner committed:
Update docs/changelog/128043.yaml

DaveCTurner committed:
Revert "Update docs/changelog/128043.yaml"

This reverts commit 81cef8c5b4098808d41a3a79c91eb3d314b954d0.

DaveCTurner committed:
Update docs/changelog/128043.yaml

elasticsearchmachine commented:
Pinging @elastic/es-distributed-coordination (Team:Distributed Coordination)

elasticsearchmachine commented:
Hi @DaveCTurner, I've updated the changelog YAML for you. *Note* that since this PR is labelled `>breaking`, you need to update the changelog YAML to fill out the extended information sections.

DaveCTurner committed:
Revert "Update docs/changelog/128043.yaml"

This reverts commit 6cca2b355b6d3efaa73c8abc62348e15c14756a6.

DaveCTurner committed:
Update docs/changelog/128043.yaml

elasticsearchmachine commented:
Hi @DaveCTurner, I've updated the changelog YAML for you. *Note* that since this PR is labelled `>breaking`, you need to update the changelog YAML to fill out the extended information sections.

DaveCTurner committed:
Revert "Update docs/changelog/128043.yaml"

This reverts commit b9da53eaecfa660c7f9aa3952d00cea7e90876dc.

ywangd commented on modules/repository-s3/src/test/java/org/elasticsearch/repositories/s3/AddPurposeCustomQueryParameterTests.java:
```diff
@@ -0,0 +1,155 @@
+
+            @SuppressForbidden(reason = "this test uses a HttpServer to emulate an S3 endpoint")
+            @Override
+            protected HttpHandler createHandler() {
+                return new AssertingHandler();
+            }
```
Should we make `AssertingHandler` delegate to the handler returned by `super.createHandler()` so that we don't miss the fix from #102976?

DaveCTurner replied:
Sure see 6b91e9b8e0f39a4066db4a22323faa8e614bf68a. It doesn't really matter here, the test fails either way.

ywangd commented on modules/repository-s3/src/test/java/org/elasticsearch/repositories/s3/AddPurposeCustomQueryParameterTests.java:
```diff
@@ -0,0 +1,155 @@
+                return new AssertingHandler();
+            }
+        };
+        httpFixture.apply(new Statement() {
```
Do we need to stop the fixture explicitly especially when the test fails?

DaveCTurner replied:
That's handled within `apply()`.

ywangd commented on x-pack/plugin/snapshot-repo-test-kit/qa/s3/src/javaRestTest/java/org/elasticsearch/repositories/blobstore/testkit/analyze/S3RepositoryAnalysisRestIT.java:
```diff
@@ -33,7 +38,39 @@ public class S3RepositoryAnalysisRestIT extends AbstractRepositoryAnalysisRestTe
+                return;
+            }
+        }
+        repoAnalysisStarted.set(true);
```
I think it's more readable to move this statement to be the last statement inside the `if (repoAnalysisStarted.get() == false)` block.

DaveCTurner replied:
Ok see 0e58906bf915d2b535c38bd88449ac734fca1fe1

ywangd commented on x-pack/plugin/snapshot-repo-test-kit/qa/s3/src/javaRestTest/java/org/elasticsearch/repositories/blobstore/testkit/analyze/S3RepositoryAnalysisRestIT.java:
```diff
@@ -33,7 +38,39 @@ public class S3RepositoryAnalysisRestIT extends AbstractRepositoryAnalysisRestTe
+        if (request.path().startsWith("/bucket/base_path_integration_tests/temp-analysis-")) {
+            repoAnalysisStarted.set(true);
+        }
+        if (repoAnalysisStarted.get() == false) {
```
With #126593, the s3 fixture now runs with multiple threads. Is there any racing concern with checking the atomic boolean? I guess it's probably not an issue since operations before repo analysis are all sequential?

DaveCTurner replied:
All the earlier operations relate to registering the repository and complete strictly before the start of repo analysis, regardless of concurrency in the handler.

ywangd reviewed (approved):
LGTM

DaveCTurner committed:
Merge branch 'main' into 2025/05/12/optional-s3-custom-query-parameter

DaveCTurner committed:
Rethrow errors properly

DaveCTurner committed:
No need to set `repoAnalysisStarted` again

DaveCTurner committed:
Update docs/changelog/128043.yaml

elasticsearchmachine commented:
Hi @DaveCTurner, I've updated the changelog YAML for you. *Note* that since this PR is labelled `>breaking`, you need to update the changelog YAML to fill out the extended information sections.

DaveCTurner committed:
Revert "Update docs/changelog/128043.yaml"

This reverts commit f38515d4fde6644ab2360cb4a95492ac6fdf454a.

DaveCTurner commented:
Backports:

- `9.0` at https://github.com/elastic/elasticsearch/pull/128178
- `8.19` at https://github.com/elastic/elasticsearch/pull/128180
- `8.18` at https://github.com/elastic/elasticsearch/pull/128185
- `8.17` at https://github.com/elastic/elasticsearch/pull/128189
